src/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Regex.cs

```diff
@@ -47,12 +47,58 @@ public class Regex
 
         // *********** } match timeout fields ***********
 
+        protected internal RegexRunnerFactory factory;
 
-        internal Dictionary<Int32, Int32> _caps;            // if captures are sparse, this is the hashtable capnum->index
-        internal Dictionary<String, Int32> _capnames;       // if named captures are used, this maps names->index
+        internal Dictionary<int, int> _caps;            // if captures are sparse, this is the hashtable capnum->index
+        internal Dictionary<string, int> _capnames;     // if named captures are used, this maps names->index
 
-        internal String[] _capslist;                        // if captures are sparse or named captures are used, this is the sorted list of names
-        internal int _capsize;                              // the size of the capture array
+        protected internal String[] capslist;              // if captures are sparse or named captures are used, this is the sorted list of names
+        protected internal int capsize;                    // the size of the capture array
+        protected IDictionary Caps
+        {
+            get
+            {
+                return _caps;
+            }
+            set
+            {
+                if (value is Dictionary<int, int>)
+                {
+                    _caps = (Dictionary<int, int>)value;
+                }
+                else
+                {
+                    _caps = new Dictionary<int, int>();
+                    foreach(DictionaryEntry entry in value)
+                    {
+                        _caps.Add((int)entry.Key, (int)entry.Value);
+                    }
+                }
```

Comment:
If it's unlikely this fallback path is ever taken in practice (i.e. if the tool-generated code is always going to use `Dictionary<int, int>`), then the extra box allocations don't matter that much and `foreach` is ok.

If it were me, I'd still do the cast once and specify the capacity, though.

```csharp
if (value == null)
{
    throw new ArgumentNullException(nameof(value));
}

_caps = value as Dictionary<int, int>;
if (_caps == null)
{
    _caps = new Dictionary<int, int>(value.Count);
    foreach (DictionaryEntry entry in value)
    {
        _caps.Add((int)entry.Key, (int)entry.Value);
    }
}
```